license-maven-plugin/src/main/java/com/mycila/maven/plugin/license/AbstractLicenseMojo.java

```diff
@@ -72,26 +63,35 @@
  */
 public abstract class AbstractLicenseMojo extends AbstractMojo {
 
+    @Parameter
+    public LicenseSet[] licenseSets;
+
     /**
      * The base directory, in which to search for project files.
      */
-    @Parameter(property = "license.basedir", defaultValue = "${basedir}", required = true)
-    public File basedir;
+    @Parameter(property = "license.basedir", defaultValue = "${basedir}", alias = "basedir", required = true)
+    public File baseBasedir;
```

Comment:
why baseBasedir ? could we have a better name ?

Reply:
So the `base` prefix is used to imply that this value is the base value, i.e. it can be overriden in `licenseSet`. If its not provided in a `licenseSet` then the licenseSet uses the base value ;-)

Reply:
ok!
And what happens if some "licenseSet" definitions have intersections ? This is eagearly detected ? Or the formatter / checker will run on the same files with different licenses ?

Reply:
I am not quite sure what you mean by "intersections"?  ....The entire logic that was there previously, is now run once per-licenseSet.

Reply:
If I understand correctly, `licenseSets` is equivalent to sequence of a plugin definition (headers + include/excludes).

So 2 questions:

1. What if the user defines different kind of licenses but turns out some set have the same inclusions ?
2. What's the different with defining multiple maven execution sets, each one with its own configuration and with different ids ?

```xml
<plugin>
    <groupId>com.mycila</groupId>
    <artifactId>license-maven-plugin</artifactId>
    <version>X.Y.ga</version>
    <configuration>
[global config here]
    </configuration>
    <executions>
        <execution>
            <id>first</id>
            <goals>
                <goal>check</goal>
            </goals>
            <configuration>
[config here for first run with its own includes]
            </configuration>
        </execution>
        <execution>
            <id>secondid>
            <goals>
                <goal>check</goal>
            </goals>
            <configuration>
[config here for second run with its own includes]
            </configuration>
        </execution>
    </executions>
</plugin>
```

Reply:
> What if the user defines different kind of licenses but turns out some set have the same inclusions ?

As set will operate on whatever inclusions it has. So if two sets share inclusions, both will be applied to the same included files.

> What's the different with defining multiple maven execution sets, each one with its own configuration and with different ids ?

Ahhhh... Glad you asked! So there is a longer story behind my motivation. Originally I was using multiple execution.... I added this PR because I was told by the Maven maintainers that using such executions in that manner was not a good idea, and that I should improve the plugin itself - see this PR to Maven for background - https://github.com/apache/maven/pull/357

Reply:
How is it a good idea ? Maven is clearly allowing this: to have the same plugin executed several times on different execution ids and eventually phases.  What use case you cannot solve by using that plus eventually profiles ?

Reply:
> Maven is clearly allowing this

Yes I agree, it does allow that.

I don't really understand the reasoning of the response I got from the Maven project, but the advice there was that I should not use multiple executions for what I am doing, but should instead enhance the plugin to perform operations on sets of things.

Hopefully this feature is acceptable for you?

Reply:
Well, If we have the choice between solving the use case using what Maven provides and solving it by modifying and deprecating the plugin to mimic Maven's behavior, I prefer the former.
The point is not that the code is not acceptable, the point is that it solves a problens that shouldn't be solved as part of a plugin.
Tnis use case is mainly to be able to run multiple times an execution of a plugin. There is no Maven plugin that I know of that has this execution set mechanism coded in itself, especially because Maven is done to allow people to provided their own configuration inheritance with plugin configuration and different executions that can be configured too.
Where did you understand that the Maven team does not encourage to use multiple executions ? I didn't see such thing...  It is part of the inheritance mechanism of Maven also.

Reply:
> There is no Maven plugin that I know of that has this execution set mechanism coded in itsel

I just did a quick survey of one large Maven project here and it seems that there are actually lots of Maven plugins that allow you to do set based operations. I just found these:

1. maven-assembly-plugin takes a set of descriptors.
2. appassembler-maven-plugin takes a set of daemons.
3. copy-maven-plugin takes a set of resources.
4. maven-clean-plugin takes a set of filesets.
5. xml-maven-plugin takes a set of transformsets.
6. copy-rename-maven-plugin takes a set of filesets.
7. maven-dependency-plugin takes a set of artifactItem.

> Where did you understand that the Maven team does not encourage to use multiple executions

I got that impression from the feedback I was given here - https://github.com/apache/maven/pull/357

Reply:
Yes you are right!
Some are just filesets, equivalent to what we have in MLP.  The closest would be the assembly plugin...
But you are right that from a usage perspective this is fare more friendly to provide that...
You convinced me ;-)